experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -127,63 +128,86 @@ static void markPathToCommits(llvm::DenseSet<Operation *> &markedPath,
-    // To do so, a structure of two connected branches is created.
-    // A speculating branch first discards the condition in case that
-    // the data is not speculative. In case it is speculative, a new branch
-    // is created that replicates the current branch.
-
-    MLIRContext *ctx = &getContext();
-    OpBuilder builder(ctx);
-    builder.setInsertionPointAfterValue(ctrlSignal);
-
-    // The speculating branch will discard the branch's condition token if the
-    // branch output is non-speculative. Speculative tag of the token is
-    // currently implicit, so the branch input itself is used at the IR level.
-    auto branchDiscardNonSpec = builder.create<handshake::SpeculatingBranchOp>(
-        branchOp.getLoc(), currOpOperand.get() /* specTag */,
-        branchOp.getConditionOperand());
-    inheritBB(specOp, branchDiscardNonSpec);
-
-    // The replicated branch directs the control token based on the path the
-    // speculative token took
-    auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
-        branchDiscardNonSpec->getLoc(),
-        branchDiscardNonSpec.getTrueResult() /* condition */,
-        ctrlSignal /* data */);
-    inheritBB(specOp, branchReplicated);
-
-    // Follow the two branch results with a different control signal
-    for (unsigned i = 0; i <= 1; ++i) {
-      for (OpOperand &dstOpOperand : branchOp->getResult(i).getUses()) {
-        Value ctrl = branchReplicated->getResult(i);
-        routeCommitControl(markedPath, ctrl, dstOpOperand);
+void HandshakeSpeculationPass::routeCommitControl(llvm::DenseSet<Operation *> &markedPath) {
+  // Perform BFS
+  std::queue<std::tuple<const OpOperand &, CommitBranchList>> queue;
+  for (OpOperand &succOpOperand : specOp.getDataOut().getUses()) {
+    queue.push(std::tuple<const OpOperand &, CommitBranchList>(succOpOperand, {}));
+  }
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+  while (!queue.empty()) {
+    auto [currOpOperand, commitBranchList] = queue.front();
+    queue.pop();
+    Operation *currOp = currOpOperand.getOwner();
+
+    if (auto commitOp = dyn_cast<handshake::SpecCommitOp>(currOp)) {
+      // We only replicate branches only if the traverse reaches a commit.
+      // Because sometimes the commit unit is already connected to the shortest (appropriate) path
+      // when we traverse other branches.
+      Value ctrlSignal = specOp.getCommitCtrl();
+      for (auto [valueForSpecTag, branchOp, branchDir] : commitBranchList) {
+        // Replicate a branch in the control path and use new control signal.
+        // To do so, a structure of two connected branches is created.
+        // A speculating branch first discards the condition in case that
+        // the data is not speculative. In case it is speculative, a new branch
+        // is created that replicates the current branch.
+        builder.setInsertionPointAfterValue(ctrlSignal);
```

Comment:
this comment should not touch this line of code if it not specifically about this line of code